### ISSUE TITLE
[REVIEW]: Pooch: A friend to fetch your data files

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1943]
**Submitting author:** @leouieda (<a href="http://orcid.org/0000-0001-6123-9515">Uieda, L</a>)
**Repository:** <a href="https://github.com/fatiando/pooch" target ="_blank">https://github.com/fatiando/pooch</a>
**Version:** v0.7.1
**Editor:** @danielskatz
**Reviewer:** @hmaarrfk, @martindurant
**Archive:** <a href="https://doi.org/10.5281/zenodo.3611376" target="_blank">10.5281/zenodo.3611376</a>

## Status

[![status](https://joss.theoj.org/papers/8053e31d2cf8a83007fecbf7b6684e3b/status.svg)](https://joss.theoj.org/papers/8053e31d2cf8a83007fecbf7b6684e3b)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/8053e31d2cf8a83007fecbf7b6684e3b"><img src="https://joss.theoj.org/papers/8053e31d2cf8a83007fecbf7b6684e3b/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/8053e31d2cf8a83007fecbf7b6684e3b/status.svg)](https://joss.theoj.org/papers/8053e31d2cf8a83007fecbf7b6684e3b)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/fatiando/pooch" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@hmaarrfk & @martindurant, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @danielskatz know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/fatiando/pooch">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@leouieda) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/fatiando/pooch">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@leouieda) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @hmaarrfk, @martindurant it looks like you're currently assigned to review this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.7717/peerj.453 is OK
- 10.5065/D6WW7G29 is OK
- 10.21105/joss.00957 is OK
- 10.5281/ZENODO.3086002 is OK
- 10.5281/ZENODO.3542092 is OK
- 10.5281/ZENODO.1450596 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01943/joss.01943/10.21105.joss.01943.pdf)

**Comment by danielskatz** [comment]
👋 @hmaarrfk & @martindurant - here we are in the review thread

As I think you know you, your job now is go through the paper and repository and check off boxes in your checklist

Also please be sure to read the prior comments in this issue

**Comment by danielskatz** [comment]
Also, thanks for agreeing to review!

**Comment by martindurant** [comment]
What is the usual timeframe for this process?

**Comment by danielskatz** [comment]
normally a couple of weeks, but with the December holidays, a bit longer is fine

**Comment by martindurant** [comment]
Note that the links are not clickable in the github-rendered version of the PDF, only in the download. This is doubtless not a fault of the authors.

I find that the paper does indeed convey that Pooch exists, is and says how and when it might be used. It completes the purpose, therefore, of submitting to JOSS. I am still disappointed by the comparison with the rest of the field and Intake in particular. Obviously, again, I am biased here and perhaps some of the points that follow have more to do with Intake's documentation than with Pooch's, and I do not insist on any changes. However:
- the library fsspec contains local-file-cache implementations that could be seen as implementing what Pooch does, also by changing only a single line of code. I would argue that it is indeed more flexible, having many file-system backends and allowing for partial downloads of large files. It can also check the source for changes on backends that support checksum or at least write-date metadata, or general cache expiry.
- Intake does *not* require you to rewrite any code at all in the simplest case such as the pandas example, but only change one line (pd.read_csv -> intake.open_csv; same arguments) with or without local caching
- Intake does many other things too, such as read from sources which are not files, from remote servers and services; metadata and visualisation and data-source browsing. In this, I agree that Pooch does one job and does it well. 
- Indeed, Intake and Pooch should have integration points, and I think this is the biggest and saddest missing point. The Pooch registry (which files) and an Intake catalog (how to load which files) clearly have overlap. Pooch could be one way of obtaining files for Intake. 
- At the very least, Pooch may want to investigate fsspec and its backends, to enable downloading from a variety of file stores, not just HTTP/FTP. https://github.com/intake/filesystem_spec/blob/master/fsspec/registry.py#L12

**Comment by hmaarrfk** [comment]
Looks good. I think a few numbers and facts can be given to accentuate the impact of the paper.

* Scientific software is also used to "Capture" or "Acquire" data
* "The usual" -> "One common approach" when talking about smaller datasets.
* The sentence that describes the challenge with including datasets in the github repo should come before the "Larger datasets [...]" sentence.
* HTTP -> HTTPS
* You state that people are recreating the same code, but maybe give a few examples?


Paragraph 2:
* [ ] It isn't clear what the registery holds: local file name, hash, remote file url should be clearly listed.
* [ ] I "minimal dependencies" is the selling point to packagers. The selling point to users is that it is easily installable with standard python distributions PyPi and Conda, and a wide array of python versions (2.7, 3.5->3.8).

Paragraph 3:
* [ ] the functionality to unzip files is understated and should be made more explicit. It probably deserves a paragraph of its own.

Paragraph 4:
* [ ] You should elaborate on what exactly makes intake harder to setup. It seems like Pooch is "files only", while intake is "files + metadata" which is a larger hurdle to overcome.

Paragraph 5:
* [ ] I think you can show off a little bit explaining the motivation in terms of install size this would have on scikit-image. The numbers are listed in the PR

**Comment by hmaarrfk** [comment]
Truthfully, if pinged, I will accept for JOSS, but I think a few extra words could go a long way.

**Comment by danielskatz** [comment]
> Note that the links are not clickable in the github-rendered version of the PDF, only in the download. This is doubtless not a fault of the authors.

Yes, this is a function of GitHub - and some of the links in the paper will not fully work until the paper is closer to being accepted, such as the link to the archived software.

**Comment by danielskatz** [comment]
Thanks @martindurant & @hmaarrfk!

It looks to me like a bit more work would really help this paper, particularly in terms of the state of the field part.

Also, I notice both of you have not checked off the functionality box - Can you let us know what you think is still needed to enable you to check this off?

**Comment by hmaarrfk** [comment]
I would like a checkbox for `Functionality` to be incldued in the `Software Paper` section. My comments mostly refer to that.

**Comment by danielskatz** [comment]
I'm confused by your comment.

The software paper section includes

```Summary: Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?```

This is intended to ask if the paper sufficiently describes the functionality of the software at a high-level.

The functionality section includes

```Functionality: Have the functional claims of the software been confirmed?```

This is intended to ask if the software does what it says it does.

Can you explain what you would like again?

**Comment by hmaarrfk** [comment]
I guess the question is:

Do you want a deeper discussion of the functionality in the paper for a more technical audience as well?

If so, some of the comments I made refer to that.

**Comment by danielskatz** [comment]
JOSS papers are not intended to be long, and they are really intended to support/introduce the software itself and its documentation, so the paper should be more of a high-level introduction to what the software does, and the software documentation should include the deeper discussion

**Comment by danielskatz** [comment]
@hmaarrfk - To continue our discussion, if the functionality of the software is as claimed, please check off the functionality box. If you have further thoughts on the software paper or documentation, let's talk about them as well.

**Comment by danielskatz** [comment]
👋 @leouieda - I think we are also waiting for you to address some of @martindurant's comments on the state of the field description in the paper. Please let us know how this is going.

**Comment by martindurant** [comment]
(just to reiterate: my comments were friendly suggestions only, and if the authors decide not to change anything, they are welcome to say so)

**Comment by leouieda** [comment]
:wave: @danielskatz just catching up with notifications from the holidays and will start working on replies/fixes. 

@martindurant @hmaarrfk thank you for the reviews! They are all appreciated and I'll provide detailed replies soon.

> (just to reiterate: my comments were friendly suggestions only, and if the authors decide not to change anything, they are welcome to say so)

@martindurant your comments are more than welcome and that is precisely why I suggested you as a reviewer :slightly_smiling_face:

**Comment by leouieda** [comment]
> the library fsspec contains local-file-cache implementations that could be seen as implementing what Pooch does, also by changing only a single line of code. I would argue that it is indeed more flexible, having many file-system backends and allowing for partial downloads of large files. It can also check the source for changes on backends that support checksum or at least write-date metadata, or general cache expiry.

@martindurant I had seen fsspec while investigating Intake but to be honest, I didn't understand that this is what it does after reading the README or the documentation front page. In hind sight, I should have dug deeper as using fsspec would have simplified some of the work we did on Pooch.

You're right that fsspec is a much more flexible and comprehensive tool and people should definitely use it if they need this flexibility. What Pooch offers is a more limited scope and because of that it's probably easier to understand for non-specialists.

I'll include a section about fsspec in the paper as well.

> Intake does *not* require you to rewrite any code at all in the simplest case such as the pandas example, but only change one line (pd.read_csv -> intake.open_csv; same arguments) with or without local caching

Right, that is what we meant but didn't do a good job in conveying it in the text. Sorry about that. I'll revise this part to make it clear that you only need this for non-standard formats.

> Intake does many other things too, such as read from sources which are not files, from remote servers and services; metadata and visualisation and data-source browsing. In this, I agree that Pooch does one job and does it well.

I'll make sure to mention more of this in the text. We should probably also include something like this in the documentation pointing people to Intake if they need these other features.

> Indeed, Intake and Pooch should have integration points, and I think this is the biggest and saddest missing point. The Pooch registry (which files) and an Intake catalog (how to load which files) clearly have overlap. Pooch could be one way of obtaining files for Intake.

I completely agree. This has come down to a lack of time on my part to invest in really learning Intake. I'm happy to keep the door open for collaboration but given the time constraints it can't be a priority of mine (but that doesn't mean someone else couldn't take this on).

> At the very least, Pooch may want to investigate fsspec and its backends, to enable downloading from a variety of file stores, not just HTTP/FTP. https://github.com/intake/filesystem_spec/blob/master/fsspec/registry.py#L12

Right now, Pooch is pretty much a finished product since it solves the problem we initially set out to solve. But it would make a lot of sense to replace our download code with fsspec in the future and use it as our backend. That way, we can contribute back new backends to fsspec instead.

**Comment by leouieda** [comment]
Hopefully all reviewer comments have been addressed in fatiando/pooch#132 and fatiando/pooch#134

**Comment by leouieda** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01943/joss.01943/10.21105.joss.01943.pdf)

**Comment by danielskatz** [comment]
👋 @martindurant & @hmaarrfk - please let us know if you think this is now ready to accept or if anything else is needed. (@martindurant - specifically, there is one review criterion that you have not checked)

**Comment by martindurant** [comment]
yes, all go

**Comment by danielskatz** [comment]
@hmaarrfk - your opinion please?

**Comment by hmaarrfk** [comment]
Just waking up. Good on my end. I was following up in their repo and tracking there progress there :+1:

**Comment by danielskatz** [comment]
👋 @leouieda - At this point could you:
- [ ] Make a tagged release of your software, and list the version tag of the archived version here.
- [ ] Archive the reviewed software in Zenodo or a similar service (e.g. figshare, an institutional repository)
- [ ] Check the archival deposit (e.g., in Zenodo) has the correct metadata, this includes the title (should match the paper title) and author list (make sure the list is correct and people who only made a small fix are not on it); you may also add the authors' ORCID.
- [ ] Please list the DOI of the archived version here.

I can then move forward with accepting the submission. Please also proofread the submission, including the references, which I will also do.

**Comment by leouieda** [comment]
@danielskatz will do. Just waiting on any editorial fixes to the paper before releasing and archiving.

**Comment by danielskatz** [comment]
The only things I see are potentially that some of the "python"s in reference titles could be "Python"s instead.

**Comment by leouieda** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01943/joss.01943/10.21105.joss.01943.pdf)

**Comment by leouieda** [comment]
Should be fixed now. The bibtex was poorly formatted.

**Comment by leouieda** [comment]
@danielskatz  done. Release version 0.7.1 and archive DOI [10.5281/zenodo.3611376](https://doi.org/10.5281/zenodo.3611376)

**Comment by danielskatz** [comment]
@whedon set v0.7.1 as version

**Comment by whedon** [comment]
OK. v0.7.1 is the version.

**Comment by danielskatz** [comment]
@whedon set 10.5281/zenodo.3611376 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3611376" target="_blank">10.5281/zenodo.3611376</a> is the archive.

**Comment by danielskatz** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.7717/peerj.453 is OK
- 10.5065/D6WW7G29 is OK
- 10.21105/joss.00957 is OK
- 10.5281/ZENODO.3086002 is OK
- 10.5281/ZENODO.3542092 is OK
- 10.5281/ZENODO.1450596 is OK
- 10.21105/joss.01450 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1234

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1234, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by danielskatz** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1218208453728067589) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1235
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.01943](https://doi.org/10.21105/joss.01943)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by danielskatz** [comment]
Thanks to @hmaarrfk & @martindurant for editing!
And congratulations to @leouieda and co-authors!

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.01943/status.svg)](https://doi.org/10.21105/joss.01943)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01943">
  <img src="https://joss.theoj.org/papers/10.21105/joss.01943/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.01943/status.svg
   :target: https://doi.org/10.21105/joss.01943
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.01943/status.svg)](https://doi.org/10.21105/joss.01943)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by leouieda** [comment]
Thank you @danielskatz @hmaarrfk & @martindurant for all your work!
